service/group0_state_broadcaster.cc

```diff
@@ -0,0 +1,113 @@
+/*
+ * Modified by ScyllaDB
+ * Copyright 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: (AGPL-3.0-or-later and Apache-2.0)
+ */
+
+#include "group0_state_broadcaster.hh"
+#include "compaction/compaction_manager.hh"
+#include "replica/database.hh"
+
+namespace service {
+
+namespace {
+
+logging::logger slogger("group0_state_broadcaster");
+
+} // namespace
+
+void group0_state_broadcaster::start() {
+    SCYLLA_ASSERT(this_shard_id() == 0);
+
+    if (_stopped) {
+        return;
+    }
+
+    _done = make_ready_future<>();
+
+    _timer.set_callback([this] {
+        slogger.debug("Disseminating last state id ...");
+
+        _done = sys_ks()
+                    .get_last_group0_state_id()
+                    .then([this](const utils::UUID& state_id) {
+                        return gossiper().add_local_application_state(gms::application_state::STATE_ID,
+                                                                      gms::versioned_value::state_id(service::state_id(state_id)));
+                    })
+                    .then([this] {
+                        utils::UUID min_state_id;
+                        gossiper().for_each_endpoint_state([this, &min_state_id](const gms::inet_address& addr, const gms::endpoint_state& eps) {
```

Comment:
Is it guaranteed that gossip endpoint states always contain at least all group0 members?

Reply:
It better be. We take id to ip mapping from there.

Reply:
> I remember seeing in logs warnings about missing id->ip mapping for a while in some cases, then it converged after a while.

Right. During bootstrap we wait for id->ip mapping to propagate through the gossiper before continuing with bootstrap. IIRC it is done before we add the node into the raft. (Need to check the code. @kbr-scylla do you remember?)

> 
> Wouldn't it be safer to look at group0 members here? If we bootstrap a new node, at least a quorum should observe it before we start streaming to it.

Streaming what?

Here we need to check that the node we are looking at is a member of group0 (otherwise some nodes that a waiting to be bootstrapped will prevent us from collecting garbage since they will distribute minimal group state id. I do not see that we have protection against it in the code. May be I missed it).

Reply:
> > I remember seeing in logs warnings about missing id->ip mapping for a while in some cases, then it converged after a while.
> 
> Right. During bootstrap we wait for id->ip mapping to propagate through the gossiper before continuing with bootstrap. IIRC it is done before we add the node into the raft. (Need to check the code. @kbr-scylla do you remember?)
> 
> > Wouldn't it be safer to look at group0 members here? If we bootstrap a new node, at least a quorum should observe it before we start streaming to it.
> 
> Streaming what?

group0 snapshot

Consider this scenario:

1) group0 members: {A, B}, leader on A
2) bootstrap of C starts, snapshot transfer starts, node C receives and applies mutation with row R
3) leader A commits a command cmd1 which deletes row R, cmd1 is stored and applied only on {A, B}
4) on A we still see {A, B} in gossip, decide that cmd1 is repaired and we garbage-collect it in storage
5) node A makes a snapshot and truncates the raft log
6) node C restarts snapshot transfer, doesn't get deletion of row R, so row R is resurrected

Reply:
> on A we still see {A, B} in gossip, decide that cmd1 is repaired and we garbage-collect it in storage

If A does not see C in the gossip it will not know its IP and will not be able to send a snapshot to it. A leader waits for id->ip resolution for joining node before adding it to group0. It also make all other nodes in the cluster to wait. See `raft_topology_cmd::command::wait_for_ip`.

But I would change the code to go over group0 members and look each one in gossiper individually indeed. Such logic would match the intention closer.

Reply:
> > on A we still see {A, B} in gossip, decide that cmd1 is repaired and we garbage-collect it in storage
> 
> If A does not see C in the gossip it will not know its IP and will not be able to send a snapshot to it. A leader waits for id->ip resolution for joining node before adding it to group0. It also make all other nodes in the cluster to wait. See `raft_topology_cmd::command::wait_for_ip`.

Ok. But isn't it relying on implementation detail? I imagine that transfer could be implemented in a way in which only node C has to know the ip of node A.

Reply:
> > > on A we still see {A, B} in gossip, decide that cmd1 is repaired and we garbage-collect it in storage
> > 
> > 
> > If A does not see C in the gossip it will not know its IP and will not be able to send a snapshot to it. A leader waits for id->ip resolution for joining node before adding it to group0. It also make all other nodes in the cluster to wait. See `raft_topology_cmd::command::wait_for_ip`.
> 
> Ok. But isn't it relying on implementation detail? I imagine that transfer could be implemented in a way in which only node C has to know the ip of node A.

It is relaying on the design. Waiting is not just implementation detail. Successful id->ip  resolution is also needed for streaming to succeed.

Reply:
> It is relaying on the design. Waiting is not just implementation detail. Successful id->ip resolution is also needed for streaming to succeed.

Ok, then my scenario is invalid.

I wonder though if address map is guaranteed to be consistent with gossip endpoint states. There was an idea to populate address map when there is an incoming connection from a node, based on its advertised host id. In that case wait_for_ip() may succeed even if gossip still doesn't have the entry for the node.

Reply:
> > It is relaying on the design. Waiting is not just implementation detail. Successful id->ip resolution is also > 
> I wonder though if address map is guaranteed to be consistent with gossip endpoint states. There was an idea to populate address map when there is an incoming connection from a node, based on its advertised host id. In that case wait_for_ip() may succeed even if gossip still doesn't have the entry for the node.

Well yes, if we will not use gossiper to populate address map then there will be no such guaranty.